// @llamaindex/core/tools/dist/index.js
import { consoleLogger } from '@llamaindex/env';
import { z } from 'zod';
import { zodToJsonSchema } from 'zod-to-json-schema';

function isZodSchema(obj) {
    return obj !== null && typeof obj === "object" && "parse" in obj && typeof obj.parse === "function" && "safeParse" in obj && typeof obj.safeParse === "function" && "_def" in obj;
}

class FunctionTool {
    #fn;
    #additionalArg;
    #metadata;
    #zodType;
    #logger;
    constructor(fn, metadata, zodType, additionalArg, logger){
        this.#zodType = null;
        this.bind = (additionalArg)=>{
            return new FunctionTool(this.#fn, this.#metadata, this.#zodType ?? undefined, additionalArg);
        };
        this.call = (input)=>{
            let params = input;
            if (this.#zodType) {
                const result = this.#zodType.safeParse(input);
                if (result.success) {
                    params = result.data;
                } else {
                    this.#logger.warn(result.error.errors);
                }
            }
            return this.#fn.call(null, params, this.#additionalArg);
        };
        this.#fn = fn;
        this.#metadata = metadata;
        if (zodType) {
            this.#zodType = zodType;
        }
        this.#additionalArg = additionalArg;
        this.#logger = logger ?? consoleLogger;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static from(fnOrConfig, schema) {
        // Handle the case where an object with execute function is passed
        if (typeof schema === "undefined" && typeof fnOrConfig === "object" && fnOrConfig.execute) {
            const { execute, parameters, ...restConfig } = fnOrConfig;
            if (isZodSchema(parameters)) {
                const jsonSchema = zodToJsonSchema(parameters);
                return new FunctionTool(execute, {
                    ...restConfig,
                    parameters: jsonSchema
                }, parameters);
            }
            return new FunctionTool(execute, fnOrConfig);
        }
        // Handle the original cases
        if (schema && schema.parameters instanceof z.ZodSchema) {
            const jsonSchema = zodToJsonSchema(schema.parameters);
            return new FunctionTool(fnOrConfig, {
                ...schema,
                parameters: jsonSchema
            }, schema.parameters);
        }
        return new FunctionTool(fnOrConfig, schema);
    }
    get metadata() {
        return this.#metadata;
    }
}
/**
 * A simpler alias for creating a FunctionTool.
 */ const tool = FunctionTool.from;

export { FunctionTool, tool };
